### ISSUE TITLE
Unreal Editor Crashes on Stop

### ISSUE THREAD

**Issue by oandreassen** [getnamo/socketio-client-ue4#32]
When clicking on Stop after testing a blueprint the entire editor freezes. The only option is to kill the process with the task manager. This occurs most of the times - making it impossible to make any progress using the socket-io plugin.

I'm using Uneal 4.14.3, socket.IO 1.7

**Comment by oandreassen** [comment]
Seems to be related to if the socket-server is responding or not (sometimes it is caught in a breakpoint)

**Comment by getnamo** [comment]
This is correct, I've seen this behavior before. It appears the connection timeout happens on the gamethread at times which is unexpected and needs to be investigated. Are you connecting manually or via default address?

**Comment by oandreassen** [comment]
It dosent happen when I manually connect (so far). it does however happen when I hit a breakpoint in the server and wait too long before releasing it (5-10 seconds).

**Comment by oandreassen** [comment]
Having a breakpoint in the socket.io server while disconnecting the client (clicking on stop) crashes the editor every time.

**Comment by getnamo** [comment]
try the latest release for 4.15 (https://github.com/getnamo/socketio-client-ue4/releases/tag/0.4.8) and let me know if you still have the same issues.

**Comment by Amorano** [comment]
The issues remain, and I am unsure if I am missing something. Your latest commits are merely include header changes? Not sure how that is going to magically fix an editor hard lock on session timeouts?

**Comment by oandreassen** [comment]
Tested using 4.15.

Setting breakpoints anywhere in the server while the game is active seems to be working fine.
Having a break point (paused execution) on the server while clicking on Stop in the editor still freezes everything.

Thank you very much for compiling for 4.15!1!

**Comment by cattomic** [comment]
Hi, I get:
Unhandled exception at 0x00007FFF541E7E5E (UE4Editor-SocketIOClient-796.dll) in UE4Editor.exe: 0xC0000005: Access violation reading location 0xFFFFFFFFFFFFFFFF.
and editor crashes when I press stop

Using 4.15 with 0.4.8 with blueprints.

**Comment by getnamo** [comment]
If I'm understanding you correctly, the server is paused via e.g. breakpoint and it cannot respond. If you then disconnect and there is no response from the server due to breakpoint you will freeze or crash on the ue4 client?

**Comment by oandreassen** [comment]
@getnamo correct.

**Comment by getnamo** [comment]
Made some fixes and a few updates, see if https://github.com/getnamo/socketio-client-ue4/releases/tag/0.5.0 finally fixes this issue :)

**Comment by oandreassen** [comment]
Works like a char efter latest release!

**Comment by kvonflotow** [comment]
Hi there, I see this issue is closed but I am still getting consistent crashes on disconnect with 0.5.2. I'm not doing anything out of the ordinary, and it crashes when disconnect is called manually or when stopping the editor. There is somewhat heavy traffic from the server to client.

**Comment by getnamo** [comment]
Will need more detailed examples of what you're doing and logs of when it's crashing (found in Saved/Logs)

**Comment by kvonflotow** [comment]
Basically I am just connecting to a socket server with ws://domain.com, which is emitting to the client every 500ms for testing. Since I see you asked the OP, I am opening the connection manually and leave autoconnect disabled. It's running in a LAN environment only on the host computer. It all seems to work fine, and then will sometimes crash on disconnect.

The log file is quite long so I just zipped the folder
[ue4_crash_logs.zip](https://github.com/getnamo/socketio-client-ue4/files/906983/ue4_crash_logs.zip)

This line seems interesting to me:
```LogModuleManager:Warning: Found module file ../../../../../../Users/vonfl/Documents/Unreal Projects/SDLive/Plugins/socketio-client-ue4/socketio-client-ue4/Binaries/Win64/UE4Editor-SIOJPlugin.dll (API version 3106830), but it was incompatible with the current engine API version (3299760). This is likely a stale module that must be recompiled.```

**Comment by getnamo** [comment]
Whoops just noticed I left a few debug log functions active in the release, although they kinda were helpful here.

Regarding 
```.. but it was incompatible with the current engine API version (3299760). This is likely a stale module that must be recompiled.```

You may try to recompile the plugin yourself. Just add a C++ class to your project and recompile. But I don't think it will fix things since I found this in the dump
![image](https://cloud.githubusercontent.com/assets/542365/24824186/f96965ac-1bfe-11e7-845a-98357b239b8e.png)

Which points at a loose pointer error somewhere. Will need to find a way to replicate this, maybe trying to do something like sending information simultaneously while disconnecting might be the key.

**Comment by MWadstein** [comment]
Glad I found this, I thought I was going crazy lol.

I was starting to get random crashing errors that I could reproduce but did not know quite what was causing it till I read this thread and thought of rolling back the plugin.

0.4.8 does not have this issue, updating to 0.5.0 or later does have this issue.

1) Having the component on any active blueprint in the scene, even if it's not connected with auto connect or using any nodes can have this issue happen.   Just having the component active on your blueprint in the scene.

2) Open the map that you are currently in that has this blueprint that has the plugin on it using the console command for opening a map (OPEN MAPNAME) and the client should crash.  This is reproducible on a packaged project only for the most part.  I couldn't get it to crash the client when doing this with PIE or standalone view in the editor *but* doing this would end up crashing my dedicated server lol...

TLDR: New project,  Make a new BP, add in the socket IO component, add the BP to the scene and save it all.  Package the project and then run the packaged build (I did windows 64bit).  The game should load up to that map with no problems.  open the console (~) and type 'open MAPNAME' so it will load up the same map you are already in and it should crash the client.

This *ONLY* happens in 0.5.0 and above, 0.4.8 does not have this issue.  Tested on 4.15.0, 4.15.1, and the 4.15 branches on the github (originally thought I had bad source...)

**Comment by getnamo** [comment]
@MWadstein awesome and descriptive, I'll try to replicate and debug this

**Comment by MWadstein** [comment]
Just an update I narrowed this down to being an issue with possessing a character/pawn in the scene that you are opening again.  I really don't have any idea why this is happening but that about all I can say for the cause.

If I have my character, which has the plugin, as the default pawn when I load up the new map I get this crashing issue.  If I have none as the default pawn, no issue.  If I have none as the default pawn, and then spawn in my character I get no issue.  If I have none, spawn in the character, then possess it causes the crash.  

Even if I spawn in a character without the component and add it in later once the character is fully in game the next time I open the map again (like above) even with the code to spawn in this component later it still crashes =(

I went back to 0.4.8 for now as I have no problems with it.  I wish I had enough experience to actually figure this out but =(

**Comment by getnamo** [comment]
candidate fix merged in: https://github.com/getnamo/socketio-client-ue4/commit/d7b3d81201c685d4f5e01e53373ecebcc3d7f83b, test the new release and see if it has been fixed for you: https://github.com/getnamo/socketio-client-ue4/releases/tag/0.6.0

**Comment by freddy-se** [comment]
Thank you for working on this issue! I was running into those kind of crashes too and was putting all my hope into your latest release :)
Sadly, it won't compile on MacOS though.
![step1](https://cloud.githubusercontent.com/assets/20209738/25806001/16dd3846-3402-11e7-9acb-40c577fa76ea.png)
![step2](https://cloud.githubusercontent.com/assets/20209738/25806000/16dc869e-3402-11e7-968b-f5432527ffa1.png)
Any hints?

**Comment by getnamo** [comment]
@freddy-se no confirmed working build of macos has been made yet, see https://github.com/getnamo/socketio-client-ue4/issues/15 for the main issues regarding that platform. Since it's not my platform someone else will have to contribute a build of the underlying library https://github.com/getnamo/socket.io-client-cpp and boost ~1.6-1.62 on mac.

**Comment by getnamo** [comment]
closing due to fix: https://github.com/getnamo/socketio-client-ue4/commit/d7b3d81201c685d4f5e01e53373ecebcc3d7f83b. One subset of this case is now tracked at: https://github.com/getnamo/socketio-client-ue4/issues/44
